cmd/bbolt/main_test.go

```diff
@@ -135,35 +136,76 @@ func TestPageCommand_Run(t *testing.T) {
 }
 
 func TestPageItemCommand_Run(t *testing.T) {
-	db := btesting.MustCreateDBWithOption(t, &bolt.Options{PageSize: 4096})
-	srcPath := db.Path()
-
-	// Insert some sample data
-	t.Log("Insert some sample data")
-	err := db.Fill([]byte("data"), 1, 100,
-		func(tx int, k int) []byte { return []byte(fmt.Sprintf("key_%d", k)) },
-		func(tx int, k int) []byte { return []byte(fmt.Sprintf("value_%d", k)) },
-	)
-	require.NoError(t, err)
-
-	defer requireDBNoChange(t, dbData(t, srcPath), srcPath)
+	testCases := []struct {
+		name          string
+		printable     bool
+		itemId        string
+		expectedKey   string
+		expectedValue string
+	}{
+		{
+			name:          "printable items",
+			printable:     true,
+			itemId:        "0",
+			expectedKey:   "key_0",
+			expectedValue: "value_0",
+		},
+		{
+			name:          "non printable items",
+			printable:     false,
+			itemId:        "0",
+			expectedKey:   hex.EncodeToString(convertInt64IntoBytes(0 + 1)),
+			expectedValue: hex.EncodeToString(convertInt64IntoBytes(0 + 2)),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			db := btesting.MustCreateDBWithOption(t, &bolt.Options{PageSize: 4096})
+			srcPath := db.Path()
+
+			t.Log("Insert some sample data")
+			err := db.Update(func(tx *bolt.Tx) error {
+				b, bErr := tx.CreateBucketIfNotExists([]byte("data"))
+				if bErr != nil {
+					return bErr
+				}
 
-	meta := readMetaPage(t, srcPath)
-	leafPageId := 0
-	for i := 2; i < int(meta.Pgid()); i++ {
-		p, _, err := guts_cli.ReadPage(srcPath, uint64(i))
-		require.NoError(t, err)
-		if p.IsLeafPage() && p.Count() > 1 {
-			leafPageId = int(p.Id())
-		}
-	}
-	require.NotEqual(t, 0, leafPageId)
+				for i := 0; i < 100; i++ {
+					if tc.printable {
+						if bErr = b.Put([]byte(fmt.Sprintf("key_%d", i)), []byte(fmt.Sprintf("value_%d", i))); bErr != nil {
+							return bErr
+						}
+					} else {
+						k, v := convertInt64IntoBytes(int64(i+1)), convertInt64IntoBytes(int64(i+2))
+						if bErr = b.Put(k, v); bErr != nil {
+							return bErr
+						}
+					}
+				}
+				return nil
+			})
+			require.NoError(t, err)
+			defer requireDBNoChange(t, dbData(t, srcPath), srcPath)
```

Comment:
i wonder, why the database should not change, since we have inserted sample data ?

Reply:
Note that purpose of the `requireDBNoChange` is to ensure the command doesn't change the db's data.

Reply:
so `dbData` read the database file that is already on-disk .. then `requireDBNoChange` will read the new data file after the test data has been inserted in the transaction above .. 

My question is, how the database should not change, while we are inserting sample data ? 

please correct me if i am wrong

Reply:
When a "defer" statement gets executed, its value and parameters to the call are evaluated as usual and saved anew, so `dbData(t, srcPath)` is executed right away. It reads current db's data. Note that the we have already finished inserting all data at the moment.

But the actual function is not invoked until the surrounding function returns. When it gets really executed, it reads the db's data again, and compare it with the previous data read in `dbData(t, srcPath)`.

Reply:
The rough workflow:
1. insert some sample data
2. read the db's data using `dbData(t, srcPath)`
3. execute the command (the main part of the test);
4. read the db's data again, and compare it with the data read at step 2.

Reply:
gotcha, makes perfect sense now :) .. thanks tonsss 🙏🏽